### DISCUSSION TITLE
Not include Imagestream tags in Deployment Config

### DISCUSSION THREAD

**Discussion by mwinzie** [eclipse/jkube#1741] [Q&A] [answered]
I am using the openshift maven plugin to generate resources oc:resource. I however would not want to have imagestream tags included. Any idea how to go about this?

**Comment by rohanKanojia** [comment, marked as answer]
Are you building container image using S2I build strategy? I think you should be able to do it via setting `jkube.openshift.imageChangeTriggers` to false

**Reply by mwinzie** [reply]
Building the docker image in kubernetes mode so am using the kubernetes maven plugin. To deploy am using openshift maven plugin because I want the deployment config, oc:resource to generate the manifest and oc:helm to create helm and install.

**Reply by rohanKanojia** [reply]
@mwinzie : I see. Is running `mvn oc:resource -Djkube.openshift.imageChangeTriggers=false` working ?

**Reply by mwinzie** [reply]
I am running mvn oc:resource -Pprofile -DskipTests -Djkube.generator.spring-boot.name=image_name
can I include it as a property in the pom file?

**Reply by rohanKanojia** [reply]
yes, you can include it in profile's property section like this: https://github.com/eclipse/jkube/blob/master/kubernetes-maven-plugin/it/src/it/namespace/pom.xml#L80

**Reply by mwinzie** [reply]
Yes, seems to be working fine, though am getting this error:
Starting the Java application using /opt/run-java/run-java.sh ...
ERROR: Neither JAVA_MAIN_CLASS nor JAVA_APP_JAR is set and 0 found in /deployments (1 expected)

**Reply by rohanKanojia** [reply]
It looks like no jar has been copied in `/deployments` folder (which run-sh script uses for java -jar startup). Is it possible to share your image configuration?

**Reply by mwinzie** [reply]
I have two profiles: one for building the image(kubernetes mode) and the second one for resource and helm(openshift)
```
<profiles>
		<profile>
			<id>openshift-deployment</id>
			<properties>
				<jkube.build.strategy>docker</jkube.build.strategy>
			</properties>
			<build>
				<plugins>
					<plugin>
						<groupId>org.eclipse.jkube</groupId>
						<artifactId>kubernetes-maven-plugin</artifactId>
						<version>${jkube.version}</version>
						<configuration>
							<images>
								<image>
									<name>##image##</name>
									<build>
									<from>registry.redhat.io/fuse7/fuse-java-openshift-jdk11-rhel8:1.10-12</from>
										<createImageOptions>
											<platform>(_shared_platform)</platform>
										</createImageOptions>
									</build>
								</image>
							</images>
							<helm>
								<chart> ${project.artifactId}</chart>
								<version> ${project.version} </version>
							</helm>
						</configuration>
						<executions>
							<execution>
								<goals>
									<goal>resource</goal>
									<goal>build</goal>
									<goal>helm</goal>
								</goals>
							</execution>
						</executions>
					</plugin>
				</plugins>
			</build>
		</profile>
<profile>
			<id>popenshift-deployment</id>
			<properties>
				<jkube.build.strategy>docker</jkube.build.strategy>
				<jkube.s2i.imageStreamLookupPolicyLocal>false</jkube.s2i.imageStreamLookupPolicyLocal>
				<jkube.openshift.imageChangeTriggers>false</jkube.openshift.imageChangeTriggers>
			</properties>
			<build>
				<plugins>
					<plugin>
						<groupId>org.eclipse.jkube</groupId>
						<artifactId>openshift-maven-plugin</artifactId>
						<version>${jkube.version}</version>
						<configuration>
							<images>
								<image>
									<name>##image##</name>
									<build>
									<from>registry.redhat.io/fuse7/fuse-java-openshift-jdk11-rhel8:1.10-12</from>
									<!--fromMode>registry.redhat.io/fuse7/fuse-java-openshift-jdk11-rhel8:1.10-12</fromMode-->
										<createImageOptions>
											<platform>(_shared_platform)</platform>
										</createImageOptions>
									</build>
								</image>
							</images>
							<helm>
								<chart> ${project.artifactId}</chart>
								<version> ${project.version} </version>
							</helm>
						</configuration>
						<executions>
							<execution>
								<goals>
									<goal>resource</goal>
									<goal>helm</goal>
								</goals>
							</execution>
						</executions>
					</plugin>
				</plugins>
			</build>
		</profile>
	</profiles>
```

**Reply by mwinzie** [reply]
I run mvn clean package before mvn oc:resource

**Reply by rohanKanojia** [reply]
You'd need to add an `<assembly>` element in your `<build>` configuration to configure what files you want to copy into image. Here is an example

```xml
                <assembly>
                  <mode>dir</mode>
                  <targetDir>/deployments</targetDir>
                  <layers>
                    <layer>
                      <id>customized-quarkus</id>
                      <files>
                        <file>
                          <source>target/quarkus-app/quarkus-run.jar</source>
                          <outputDirectory>.</outputDirectory>
                        </file>
                      </files>
                      <fileSets>
                        <fileSet>
                          <directory>${project.basedir}/target/lib</directory>
                          <outputDirectory>lib</outputDirectory>
                        </fileSet>
                      </fileSets>
                    </layer>
                  </layers>
                </assembly>
```

**Reply by mwinzie** [reply]
The assembly element gave an error.

Also will the fromMode configuration change the base image builder to this:

 Using Docker image registry.redhat.io/fuse7/fuse-java-openshift-jdk11-rhel8:1.10-12 as base / builder

![image](https://user-images.githubusercontent.com/27803878/188874859-41984317-1d4d-47a8-951d-8f0dec6adec0.png)

**Reply by rohanKanojia** [reply]
What error are you getting?

**Reply by mwinzie** [reply]
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.8.0:build (default-cli) on project: Failed to execute the build: Error while trying to build the image: /home/azureuser/myagent/_work/9/s/project/target/project -> [Help 1]

If there is a way the oc:resource task can get the base image builder to be registry.redhat.io/fuse7/fuse-java-openshift-jdk11-rhel8:1.10-12 it may solve the problem. Tried using `<from>`  and `<fromMode>` but got the error

[ERROR] Failed to execute goal org.eclipse.jkube:openshift-maven-plugin:1.8.0:resource (default-cli) on project : Unable to parse configuration of mojo org.eclipse.jkube:openshift-maven-plugin:1.8.0:resource for parameter fromMode: Cannot find 'fromMode' in class org.eclipse.jkube.kit.config.image.build.BuildConfiguration -> [Help 1]

**Reply by mwinzie** [reply]
Copied but still getting the same error:
Starting the Java application using /opt/run-java/run-java.sh ...
ERROR: Neither JAVA_MAIN_CLASS nor JAVA_APP_JAR is set and 0 found in /deployments (1 expected)

```
                <assembly>
                  <mode>dir</mode>
                  <targetDir>/deployments</targetDir>
				  <inline>
                      <fileSets>
                        <fileSet>
							<directory>${project.basedir}/target</directory>                        
							<outputDirectory>deployments/</outputDirectory>
							<includes>
								<include>*.jar</include>
						  </includes>
                        </fileSets>
                      </fileSet>
				  </inline>
                </assembly> 
```

**Reply by rohanKanojia** [reply]
Are you sure there is a jar file present in target folder? What if you specify complete file in `<directory>` element?

**Reply by mwinzie** [reply]
Yes, it's there unless am getting the path wrong. Do I need the assembly plugin?
![image](https://user-images.githubusercontent.com/27803878/189207012-04e6f6d2-4cc0-4284-b32f-99f4f4d36631.png)

Tried a different directory still not working:
```
									<mode>dir</mode>
									<targetDir>/deployments</targetDir>
										<files>
											<file>
											<source>target/xxx/-SNAPSHOT.jar</source>
											<outputDirectory>target/classes</outputDirectory>
											</file>
										</files>
```

**Reply by rohanKanojia** [reply]
Umm, no. Is it possible to share a reproducer project so that we can try this out?

**Reply by mwinzie** [reply]
I was getting the path wrong, works now. 
```

										<assembly>
										<mode>dir</mode>
										<targetDir>/deployments</targetDir>
										<layers>
											<layer>
											<id>customized</id>
											<files>
												<file>
												<source>target/xxx.jar</source>
												<outputDirectory>.</outputDirectory>
												</file>
											</files>
											</layer>
										</layers>
										</assembly> 
```

**Reply by mwinzie** [reply]
Assist with this https://github.com/eclipse/jkube/discussions/1763
